File: scripts/load-simulator.pl
#!/usr/bin/env perl
#*****************************************************************************************
# load-simulator.pl
#
# This script will restore the state of the iOS simulator
#
# Author   :  Gary Ash <gary.ash@icloud.com>
# Created  :   8-Feb-2026  2:48pm
# Modified :
#
# Copyright © 2026 By Gary Ash All rights reserved.
#*****************************************************************************************

#*****************************************************************************************
# libraries used
#*****************************************************************************************
use strict;
use warnings;
use Foundation;
use File::Copy;
use File::Find;

#*****************************************************************************************
# main line
#*****************************************************************************************
our %simulators;
our $mediaList = "";

my $HOME          = $ENV{"HOME"};
my $simulatorsLoc = "$HOME/Library/Developer/CoreSimulator/Devices";
my $media         = "$HOME/Documents/GeeDblA/Resources/Development/Apple/SimulatorBackup";

find(\&getSimulators, $simulatorsLoc);
find(\&addMedia,      $media);

`/Applications/Xcode.app/Contents/Developer/usr/bin/simctl shutdown all;/Applications/Xcode.app/Contents/Developer/usr/bin/simctl delete unavailable;/Applications/Xcode.app/Contents/Developer/usr/bin/simctl erase all`;
for my $key (keys %simulators) {
    if (system("/Applications/Xcode.app/Contents/Developer/usr/bin/simctl boot \"$key\" &> /dev/null") == 0) {
        if (system("/Applications/Xcode.app/Contents/Developer/usr/bin/simctl addmedia booted $mediaList") != 0) {
            exit(1);
        }
        if (system("/Applications/Xcode.app/Contents/Developer/usr/bin/simctl shutdown booted") != 0) {
            exit(1);
        }

    }
}
`/Applications/Xcode.app/Contents/Developer/usr/bin/simctl shutdown all`;

#*****************************************************************************************
# this routine will load a hash with names and UUIDs of the currently defined simulators
#*****************************************************************************************
sub getSimulators {
    return if ($File::Find::name !~ /\/device.plist/);
    my $plist = NSMutableDictionary->dictionaryWithContentsOfFile_($File::Find::name);
    if ($plist && $$plist) {
        my $platform = $plist->objectForKey_("runtime")->UTF8String;
        if ($platform =~ /.*iOS*/) {
            my $name = $plist->objectForKey_("name")->UTF8String;
            my $uuid = $plist->objectForKey_("UDID")->UTF8String;
            $simulators{$name} = $uuid;
        }
    }
}

#*****************************************************************************************
# this routine will build a list of all the media in the SimulatorBackup folder
#*****************************************************************************************
sub addMedia {
    return if $_ =~ /^\..*/;
    return if !-f $_;

    if ($mediaList ne "") {
        $mediaList .= " ";
    }
    $mediaList .= $File::Find::name;
}

File: scripts/load-simulator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ****************************************************************************************
#  load-simulator.py
#
# This script will restore the state of the iOS simulator
#
#  Author   :  Gary Ash <gary.ash@icloud.com>
#  Created  :   1-Sep-2026  4:42pm
#  Modified :
#
#  Copyright © 2026 By Gary Ash All rights reserved.
# ****************************************************************************************
import os
import plistlib
import subprocess
import sys
from pathlib import Path

SIMCTL = "/Applications/Xcode.app/Contents/Developer/usr/bin/simctl"
ICON_NAME = "Icon\r"


# ****************************************************************************************
# this routine will build a dictionary of names and UUIDs of the currently defined
# simulators
# ****************************************************************************************
def get_simulators(simulators_location):
    simulators = {}

    for root, _dirs, files in os.walk(simulators_location):
        if "device.plist" not in files:
            continue

        try:
            with open(Path(root) / "device.plist", "rb") as plist_file:
                plist = plistlib.load(plist_file)
        except (OSError, plistlib.InvalidFileException):
            continue

        platform = plist.get("runtime", "")
        if "iOS" in platform:
            simulators[plist["name"]] = plist["UDID"]

    return simulators


# ****************************************************************************************
# this routine will build a list of all the media in the SimulatorBackup folder
# ****************************************************************************************
def collect_media(media_location):
    media = []

    for root, _dirs, files in os.walk(media_location):
        for name in files:
            if name.startswith(".") or name == ICON_NAME:
                continue
            path = Path(root) / name
            if path.is_file():
                media.append(str(path))

    return media


# ****************************************************************************************
# main line
# ****************************************************************************************
def main():
    home = Path(os.environ["HOME"])
    simulators_location = home / "Library/Developer/CoreSimulator/Devices"
    media_location = (
        home / "Documents/GeeDblA/Resources/Development/Apple/SimulatorBackup"
    )

    simulators = get_simulators(simulators_location)
    media = collect_media(media_location)

    for arguments in (["shutdown", "all"], ["delete", "unavailable"], ["erase", "all"]):
        subprocess.run([SIMCTL] + arguments, capture_output=True, check=False)

    for name in simulators:
        booted = subprocess.run(
            [SIMCTL, "boot", name], capture_output=True, check=False
        )
        if booted.returncode != 0:
            continue

        if (
            subprocess.run(
                [SIMCTL, "addmedia", "booted"] + media, check=False
            ).returncode
            != 0
        ):
            return 1

        if subprocess.run([SIMCTL, "shutdown", "booted"], check=False).returncode != 0:
            return 1

    subprocess.run([SIMCTL, "shutdown", "all"], capture_output=True, check=False)
    return 0


if __name__ == "__main__":
    sys.exit(main())
